damnit/gui/process.py

```diff
@@ -0,0 +1,207 @@
+        current_range_end = r
+    range_ends.append(current_range_end)
+
+    s_pieces = []
+    for start, end in zip(range_starts, range_ends):
+        if start == end:
+            s_pieces.append(str(start))
+        else:
+            s_pieces.append(f"{start}-{end}")
+
+    return ", ".join(s_pieces)
+
+
+class ProcessingDialog(QtWidgets.QDialog):
+    selected_runs = ()
+    all_vars_selected = False
+    no_vars_selected = False
+
+    def __init__(self, proposal: str, runs: list[int], var_ids_titles, parent=None):
+        super().__init__(parent)
+
+        self.setWindowTitle("Process runs")
+
+        main_vbox = QtWidgets.QVBoxLayout()
+        self.setLayout(main_vbox)
+
+        hbox1 = QtWidgets.QHBoxLayout()
+        main_vbox.addLayout(hbox1)
+
+        grid1 = QtWidgets.QGridLayout()
+        hbox1.addLayout(grid1)
+        vbox2 = QtWidgets.QVBoxLayout()
+        hbox1.addLayout(vbox2)
+
+        self.edit_prop = QtWidgets.QLineEdit(proposal)
+        #self.edit_prop.setInputMask('999900;_')  # 4-6 digits
+        grid1.addWidget(QtWidgets.QLabel("Proposal:"), 0, 0)
+        grid1.addWidget(self.edit_prop, 0, 1)
+
+        self.edit_runs = QtWidgets.QLineEdit(fmt_run_ranges(runs))
+        self.edit_runs.textChanged.connect(self.validate_runs)
+        self.runs_hint = QtWidgets.QLabel(RUNS_MSG)
+        self.runs_hint.setAlignment(Qt.AlignHCenter | Qt.AlignTop)
+        self.runs_hint.setWordWrap(True)
+        grid1.addWidget(QtWidgets.QLabel("Runs:"), 1, 0)
+        grid1.addWidget(self.edit_runs, 1, 1)
+        grid1.addWidget(self.runs_hint, 2, 0, 1, 2)
+
+        self.vars_list = QtWidgets.QListWidget()
+        vbox2.addWidget(self.vars_list)
+
+        self.btn_select_all = QtWidgets.QPushButton("Select all")
+        self.btn_select_all.clicked.connect(self.select_all)
+        self.btn_deselect_all = QtWidgets.QPushButton("Deselect all")
+        self.btn_deselect_all.clicked.connect(self.deselect_all)
+        hbox_select_btns = QtWidgets.QHBoxLayout()
+        hbox_select_btns.addWidget(self.btn_select_all)
+        hbox_select_btns.addWidget(self.btn_deselect_all)
+        vbox2.addLayout(hbox_select_btns)
+
+        self.dlg_buttons = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
+        self.dlg_buttons.button(QDialogButtonBox.Ok).setEnabled(False)
+        self.dlg_buttons.accepted.connect(self.accept)
+        self.dlg_buttons.rejected.connect(self.reject)
+        main_vbox.addWidget(self.dlg_buttons)
+
+        for var_id, title in var_ids_titles:
+            itm = QtWidgets.QListWidgetItem(title)
+            itm.setData(Qt.UserRole, var_id)
+            itm.setCheckState(Qt.Unchecked if var_id in deselected_vars else Qt.Checked)
+            self.vars_list.addItem(itm)
+
+        self.vars_list.itemChanged.connect(self.validate_vars)
+
+        self.validate_runs()
+        self.validate_vars()
+
+        self.edit_runs.setFocus()
+
+    def validate_runs(self):
```

Comment:
I would probably be useful to cehck if the selected runs exist, and:
- show a warning message if they don't exist
- do not start a job for runs that do not exist
- show an error if trying to reprocess and none of the selected runs exist

Reply:
OK, how about like this...

1. All chosen runs valid:

![image](https://github.com/user-attachments/assets/c33caf03-40cd-4ff3-9d0a-6708bb8fd5b8)

2. Some don't exist:

![image](https://github.com/user-attachments/assets/c2c2f49b-389f-4dbf-be8b-0bf59e479a7f)

3. No runs exist, or the proposal number is wrong, or the user running the GUI doesn't have access:

![image](https://github.com/user-attachments/assets/70001242-ba8e-4347-b7b9-f7f915a43e9d)

In the last case, the 'Ok' button is disabled so you have to change it to launch reprocessing.

Reply:
Great!